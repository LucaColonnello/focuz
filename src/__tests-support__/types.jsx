import React, { PropTypes } from 'react';

// Container component
const Container = ({ children, className }) => (
  <div className={className}>{children}</div>);
Container.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,
};

// Header component
const Header = ({ text }) => (<h1>{text}</h1>);
Header.propTypes = {
  text: PropTypes.string,
};

// Text component
const Text = ({ text }) => (<p>{text}</p>);
Text.propTypes = {
  text: PropTypes.string,
};

// Image Component
const Image = ({ src }) => (<img alt="" src={src} />);
Image.propTypes = {
  src: PropTypes.string,
};

export default {
  Container,
  Header,
  Text,
  Image,
};
