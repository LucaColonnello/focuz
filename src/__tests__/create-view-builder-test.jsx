import renderer from 'react-test-renderer';
import createViewBuilder, { createFocuzElement } from '../';
import types from '../__tests-support__/types';

describe('createViewBuilder', () => {
  describe('render', () => {
    it('should render using react.createElement', () => {
      const viewBuilder = createViewBuilder(types);

      const component = viewBuilder(createFocuzElement(
        'Container',
        {
          className: 'main',
        },
        [
          createFocuzElement(
            'Header',
            {
              text: 'My Header',
            },
          ),
          createFocuzElement(
            'Container',
            {
              className: 'content',
            },
            [
              createFocuzElement(
                'Image',
                {
                  src: 'http://placehold.it/350x150',
                },
              ),
              createFocuzElement(
                'Text',
                {
                  text: 'Lorem Ipsum dolor sit amen...',
                },
              ),
            ],
          ),
        ],
      ));

      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should throw error when the element is not a focuz element', () => {
      const viewBuilder = createViewBuilder(types);

      expect(viewBuilder.bind(null, {
        type: 'Container',
        props: {},
      })).toThrowError(/Focuz viewBuilder: the component is not a valid element ->/);
    });
  });

  describe('modifiers', () => {
    it('should use modifiers if passed down', () => {
      const viewBuilder = createViewBuilder(types, [
        (comp, { propTypes: { className } }) => ({
          ...comp,
          props: {
            ...comp.props,
            className: className ? `${comp.props.className} modified` : undefined,
          },
        }),
      ]);

      const component = viewBuilder(createFocuzElement(
        'Container',
        {
          className: 'my-container',
        },
        [
          createFocuzElement(
            'Header',
            {
              text: 'My Header',
            },
          ),
          createFocuzElement(
            'Image',
            {
              src: 'http://placehold.it/350x150',
            },
          ),
          createFocuzElement(
            'Text',
            {
              text: 'Lorem Ipsum dolor sit amen...',
            },
          ),
        ],
      ));

      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should use different type if a modifier change it', () => {
      const viewBuilder = createViewBuilder(types, [
        comp => ({
          ...comp,
          type: comp.type === 'Header' ? 'Text' : comp.type,
        }),
      ]);

      const component = viewBuilder(createFocuzElement(
        'Header',
        {
          text: 'My Text',
        },
      ));

      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should use different type if a modifier set it to a Component', () => {
      const viewBuilder = createViewBuilder(types, [
        comp => ({
          ...comp,
          type: comp.type === 'Text' ? types.Header : comp.type,
        }),
      ]);

      const component = viewBuilder(createFocuzElement(
        'Text',
        {
          text: 'My Header',
        },
      ));

      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
