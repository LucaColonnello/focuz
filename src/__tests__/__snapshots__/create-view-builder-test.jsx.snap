exports[`createViewBuilder modifiers should use different type if a modifier change it 1`] = `
<p>
  My Text
</p>
`;

exports[`createViewBuilder modifiers should use different type if a modifier set it to a Component 1`] = `
<h1>
  My Header
</h1>
`;

exports[`createViewBuilder modifiers should use modifiers if passed down 1`] = `
<div
  className="my-container modified">
  <h1>
    My Header
  </h1>
  <img
    alt=""
    src="http://placehold.it/350x150" />
  <p>
    Lorem Ipsum dolor sit amen...
  </p>
</div>
`;

exports[`createViewBuilder render should render using react.createElement 1`] = `
<div
  className="main">
  <h1>
    My Header
  </h1>
  <div
    className="content">
    <img
      alt=""
      src="http://placehold.it/350x150" />
    <p>
      Lorem Ipsum dolor sit amen...
    </p>
  </div>
</div>
`;
